salt/_utils/volume_utils.py

```diff
@@ -0,0 +1,297 @@
+
+# blkid_probe_set_superblocks_flags
+probe_set_superblocks_flags = blkid.blkid_probe_set_superblocks_flags
+probe_set_superblocks_flags.restype  = ctypes.c_int
+probe_set_superblocks_flags.argtypes = [ctypes.c_void_p, ctypes.c_int]
+probe_set_superblocks_flags.errcheck = _check_error_code
+
+
+# Probing flags for the superblocks prober.
+SUBLKS_LABEL    = 1 <<  1  # Read LABEL from superblock.
+SUBLKS_LABELRAW = 1 <<  2  # Read and define LABEL_RAW result value.
+SUBLKS_UUID     = 1 <<  3  # Read UUID from superblock.
+SUBLKS_UUIDRAW  = 1 <<  4  # Read and define UUID_RAW result value.
+SUBLKS_TYPE     = 1 <<  5  # Define TYPE result value.
+SUBLKS_SECTYPE  = 1 <<  6  # Define compatible fs type (second type).
+SUBLKS_USAGE    = 1 <<  7  # Define USAGE result value.
+SUBLKS_VERSION  = 1 <<  8  # Read FS type from superblock.
+SUBLKS_MAGIC    = 1 <<  9  # Define SBMAGIC and SBMAGIC_OFFSET.
+SUBLKS_BADCSUM  = 1 << 10  # Allow a bad checksum.
+SUBLKS_DEFAULT  = (SUBLKS_LABEL | SUBLKS_UUID | SUBLKS_TYPE | SUBLKS_SECTYPE)
+
+
+# }}}
+# }}}
+# High level wrapper {{{
+
+
+# No enum in Python 2 :-(
+class SuperblockFlags(object):
+    """Probing flags for the superblocks prober."""
+    LABEL    = SUBLKS_LABEL
+    LABELRAW = SUBLKS_LABELRAW
+    UUID     = SUBLKS_UUID
+    UUIDRAW  = SUBLKS_UUIDRAW
+    TYPE     = SUBLKS_TYPE
+    SECTYPE  = SUBLKS_SECTYPE
+    USAGE    = SUBLKS_USAGE
+    VERSION  = SUBLKS_VERSION
+    MAGIC    = SUBLKS_MAGIC
+    BADCSUM  = SUBLKS_BADCSUM
+    DEFAULT  = SUBLKS_DEFAULT
+
+
+class PartitionFlags(object):
+    """Probing flags for the partitions prober."""
+    FORCE_GPT     = PARTS_FORCE_GPT
+    ENTRY_DETAILS = PARTS_ENTRY_DETAILS
+    MAGIC         = PARTS_MAGIC
+
+
+DeviceInfo = collections.namedtuple(
+    'DeviceInfo', ['fstype', 'uuid', 'has_partition']
+)
+
+ProbeConfig = collections.namedtuple(
+    'ProbeConfig',
+    ['enable_superblocks', 'superblocks_flags',
+     'enable_partitions',  'partitions_flags']
+)
+
+
+def _get_flags(klass, kind, default, *args):
+    """Convert a list of flag names into a bitfield."""
+    if not args:
+        return default
+    flags = []
+    for arg in args:
+        flag = getattr(klass, arg)
+        if flag is None:
+            raise ValueError('{} is not a valid flag for the {} prober'\
+                             .format(arg, kind))
+        flags.append(flag)
+    return functools.reduce(lambda x, y: x | y, flags, 0)
+
+
+class _Probe(object):
+    def __init__(
+        self, filepath,
+        use_superblocks, superblocks_flags, use_partitions, partitions_flags,
+    ):
+        """Initialize a new probe for the device at `filepath`."""
+        self._probe = new_probe_from_filename(filepath)
+        self._configure(
+            use_superblocks, superblocks_flags, use_partitions, partitions_flags
+        )
+
+    def destroy(self):
+        """Free the resources allocated by the probe."""
+        free_probe(self._probe)
```

Comment:
And now you should set `self._probe` to None (and check for it before), otherwise one can call `destroy` (which is public API!), and cause a double-free.

Reply:
I'd argue the constructor should take a `probe` which is created externally to this class code, and then create/destroy the C-level 'probe' inside the context manager code.

Reply:
Done